test/e2e/network/ingressclass.go

```diff
@@ -50,13 +50,35 @@ var _ = common.SIGDescribe("IngressClass [Feature:Ingress]", func() {
 		framework.ExpectNoError(err)
 		defer deleteIngressClass(cs, ingressClass1.Name)
 
-		ingress, err := createBasicIngress(cs, f.Namespace.Name)
-		framework.ExpectNoError(err)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		lastFailure := ""
+
+		// the admission controller may take a few seconds to observe the ingress classes
+		if err := wait.Poll(time.Second, time.Minute, func() (bool, error) {
+			lastFailure = ""
+
+			ingress, err := createBasicIngress(cs, f.Namespace.Name)
+			if err != nil {
+				lastFailure = err.Error()
+				return false, err
+			}
+			defer func() {
+				err := cs.NetworkingV1().Ingresses(ingress.Namespace).Delete(ctx, ingress.Name, metav1.DeleteOptions{})
+				framework.Logf("%v", err)
+			}()
+
+			if ingress.Spec.IngressClassName == nil {
+				lastFailure = "Expected IngressClassName to be set by Admission Controller"
+				return false, nil
+			} else if *ingress.Spec.IngressClassName != ingressClass1.Name {
+				lastFailure = fmt.Sprintf("Expected IngressClassName to be %s, got %s", ingressClass1.Name, *ingress.Spec.IngressClassName)
+				return false, nil
```

Comment:
I think we should fail directly here, if the name is set is that something is wrong

Reply:
> I think we should fail directly here, if the name is set is that something is wrong

that could simply mean that the cache in the admission plugin hasn't yet been cleared from another test.

Reply:
If there's an existing class which indicates that it is the default, and then we create a new one that thinks it is the default, retrying here will paper over that until the controller observes the new one, which is correct.  Ugly, but correct.  I think,

Reply:
ic, this is a consequence of adding these retries.
The test sets up a clean environment using a random namespace, to avoid this collision I think we can append the iteration number to the ingress name and this way we can discriminate ... I don't think we should test here the case that the same ingressclass is created multiple times

Reply:
I'm not sure I follow your comment, Antonio.  I think the issue here is something like:

* some other test creates an ingressclass A which is default
* other test completes and deletes the ingress class A
* admission control HAS NOT seen the delete yet
* this test starts and creates an ingressclass B which is default
* admission control STILL HAS NOT seen the delete yet
* this test creates an ingress which is assigned class A
* admission control finally sees the delete and create
* this test retries

Reply:
doh 

> // +genclient:nonNamespaced

I was thinking ingressClasss was namespaced all the time, forgive my comments